### DISCUSSION TITLE
How to efficiently perform connected components analysis / detecting surfaces

### DISCUSSION THREAD

**Discussion by kalsan** [libvips/libvips#3699] [Q&A] [answered]
Hello,

I've spent a considerable time getting libvips to tell me the location of red surfaces in an otherwise black-and-white image. Here is an example:

![grafik](https://github.com/libvips/libvips/assets/7618650/653dddeb-4f96-4dd5-b8e0-f44c2e27d74f)

I need to know the 4 integers of pixels (min x,y; max x,y) of the bounding box of each red surface (i.e. connected components analysis).

The documentation https://www.libvips.org/API/current/libvips-arithmetic.html#vips-hist-find-indexed states:

> This operation is useful in conjunction with [vips_labelregions()](https://www.libvips.org/API/current/libvips-morphology.html#vips-labelregions). You can use it to find the centre of gravity of blobs in an image, for example.

However, my image processing knowledge is insufficient to understand how to use `labelregions` in my use case, and I cannot find a single guide, howto, article, example or stackoverflow question that provides a detailled guide on how to achieve this.

After some time, I came up with the following Ruby code that works but is painfully slow and clearly a naive implementation:

```ruby
    input_image = Vips::Image.new_from_buffer(my_input_image, '')

    lab_image = input_image.colourspace(:lab)
    a_channel = lab_image[1].gaussblur(3)

    threshold = 50
    mask = a_channel > threshold
    components, output_items = mask.labelregions(segments: true)
    amount_output_items = output_items['segments']

    results = []

    (2..amount_output_items - 1).each do |label|
      component_mask = (components == label)
      _, x_min, y_min = component_mask.maxpos

      component_mask_flipped = component_mask.rotate(180)
      _, inv_x_max, inv_y_max = component_mask_flipped.maxpos
      x_max = input_image.width - inv_x_max
      y_max = input_image.height - inv_y_max

      results << [[x_min, y_min], [x_max, y_max]]
    end

    return results
```

My understanding is that this could be done much faster using `labelregions`. Could you provide an example on how to achieve this?

Any help is highly appreciated. Thanks in advance!

Best,
Kalsan

**Comment by jcupitt** [comment, marked as answer]
Hi @kalsan,

You could do it like this:

```ruby
#!/usr/bin/ruby

require 'vips'

image = Vips::Image.new_from_file(ARGV[0])

# flatten out any alpha
image = image.flatten if image.has_alpha?

# make a mono image which is 255 for red pixels, 0 elsewhere
mask = (image == [255, 0, 0]).bandand

# split into connected regions, with a unique label for each
labels = mask.labelregions

# make an xy image, where each pixel is its own coordinate
xy = Vips::Image.xyz labels.width, labels.height

# now ... make an indexed histogram, so eg. all xy pixels with label 
# 12 are summed into bucket 12
hist_xy = xy.hist_find_indexed labels

# hist[12][0] will be the sum of all x positions for label 12 ... we need the
# number of pixels with label 12 ... this is just a regular histogram of the
# label image
hist_labels = labels.hist_find

# so divide them to get the average x/y for each label
positions = hist_xy / hist_labels

# transform to an array and print
# label 0 is "no label", 1 is background, so start at 2
positions.to_a[0].drop(2).each_with_index do |position, i|
  puts "label #{i}: x = #{position[0]} y = #{position[1]}"
end
```

I see:

```
john@banana ~/try $ ./kalsan.rb ~/pics/red-dots.png 
label 0: x = 484.5 y = 168.0
label 1: x = 505.0 y = 168.0
label 2: x = 72.5 y = 169.0
label 3: x = 92.5 y = 169.0
label 4: x = 113.5 y = 169.0
label 5: x = 133.5 y = 169.0
label 6: x = 154.5 y = 168.5
label 7: x = 174.5 y = 168.5
...
label 305: x = 189.5 y = 564.5
label 306: x = 210.0 y = 564.5
```

Indexed histograms are useful for all kinds of things. There's a combine option which lets you pick min/max/sum, so you can find bounding boxes, major axis, etc.

**Reply by jcupitt** [reply]
(for your case I guess centroid is correct though, is that right?)

**Reply by jcupitt** [reply]
... did you find nip2 and vipsdisp? They are handy for developing things like this.

https://github.com/jcupitt/vipsdisp

You can save the labels image and view it in falsecolour, for example:

![image](https://github.com/libvips/libvips/assets/580843/3b468932-d353-4bf5-812a-85a053921e3a)

**Comment by kalsan** [comment]
Hi @jcupitt and thank you very much for your extremely helpful advice!! :-) This has advanced me a lot!

> (for your case I guess centroid is correct though, is that right?)

Unfortunately not, as I also need to detect the size of the rectangles. In my example, I made them nice and square, but that's not for granted. Here's an example with irregular rectangles:

![saalplan-simplest](https://github.com/libvips/libvips/assets/7618650/76da63e2-9236-4463-97e2-d796c9f3e590)

Best,
Kalsan

**Reply by jcupitt** [reply]
You can get the number of pixels in each region from `hist_labels`, so you could approximate your rectangles with different-sized circles (for example). Would that work?

If you must have the bounding box, call `hist_find_indexed` twice, once with `combine: :max` and once with `combine: :min`, though it'll be a little slower.

**Reply by kalsan** [reply]
Excellent, thank you very much. The speed and the accuracy of your solution are incredible!

For future reference, here's the code based on your suggestion above and your idea from here that detects the coordinates extremely quickly (miliseconds!) and reliably:

```ruby

    input_image = load_image() # Custom function, call new_from_file etc.

    # flatten out any alpha
    input_image = input_image.flatten if input_image.has_alpha?

    # In my case, pictures can be JPG (a.k.a. can be noisy) and can come from a scanner, thus I'm using a bit of filtering
    # Generate a mask of pixels that are red enough
    lab_image = input_image.colourspace(:lab)
    a_channel = lab_image[1].gaussblur(2) # this smooths noisy edges and prevents false positives in detection
    threshold = 50
    mask = a_channel > threshold

    # Split into connected regions, with a unique label for each
    labels = mask.labelregions

    # Make an xy image, where each pixel is its own coordinate
    xy = Vips::Image.xyz labels.width, labels.height

    # Make an indexed histogram, so eg. all xy pixels with label 12 are summed into bucket 12
    hist_xy_min = xy.hist_find_indexed labels, combine: :min
    hist_xy_max = xy.hist_find_indexed labels, combine: :max

    return hist_xy_min.to_a[0].zip(hist_xy_max.to_a[0]).drop(2)
```

Best,
Kalsan

**Reply by jcupitt** [reply]
Great! I'm glad it's working.

You can remove this line I think:

```
    hist_labels = labels.hist_find
```

Though it probably won't affect the speed.

**Reply by kalsan** [reply]
Indeed! Edited my answer.

Thanks for your amazing support and have a great day!

**Comment by kalsan** [comment]
BTW your solution is working great for centroid! Here's a picture of them being dotted onto the target canvas:

![canvas](https://github.com/libvips/libvips/assets/7618650/d5c91042-1e32-4139-98f9-8c7cf1f72608)
